### PR TITLE
Add express validator for use on post requests (CU-hx0jjc)

### PR DESCRIPTION
**Pull request by risatronic**
- Add express validator to use on post requests
- Updated `db.js` to use PG_DATABASE for env variable as my database name and username are different

Did not add validator to Twilio endpoint as that will be using `brave-alert-lib`, but can add or remove it from any routes as people see fit. I also wasn't sure how much I should be checking in there because only Twilio seems to be using more than just one thing pulled from the body aside from smart things but that validation happens within its own function & as mentioned Twilio validation is covered in the `brave-alert-lib` already but am very open to feedback :)

### CODE REVIEW

**Comment by trisapeace** [comment]
Task linked: [CU-hx0jjc Add Express validations to endpoints](https://app.clickup.com/t/hx0jjc)

**Comment by risatronic** [comment]
I also just noticed that there are more post routes in a different PR so maybe I should wait until that is merged and then revisit this 🤔

File: db/db.js
Comment (trisapeace) [inline db/db.js:13]: Can you take a look around to see where else `PG_USER` is being used for both? There is probably stuff in the README and maybe in `setup_postgres.sh` that need to be changed to do this too.

File: index.js
Comment (trisapeace) [inline index.js:336]: Does the validationErrors object already have a nice string to display? I wonder if we even need to put the `Bad request, parameters missing` part around it.

File: index.js
Comment (trisapeace) [inline index.js:347]: This also needs `state`, `rpm`, `distance`, `mov_f`, and `mov_s` in the `body` as used in `redis.addXeThruSensorData`.

Comment (trisapeace) [reply index.js:347]: Actually, now that I'm looking at this... if you're up for even more refactoring, it would be really nice if these route functions were the only ones to know about the request and response objects. To do this, functions that are called from them, like `redis.addXeThruSensorData(req, res)` would need to change to take in `locationid, state, rpm, distance, mov_f, mov_s` as separate parameters and not send back the `200` response.

This would be good for a bunch of reasons...
1. It would make it easier to see exactly what is being used and why it's being validated here.
2. It would make the called functions easier to test because you wouldn't need to match the structure of the request object or mock a response object
3. Would be generally good separation of concerns (i.e. why should the Redis module need to care that the data it was getting came from an API endpoint)

However, this change isn't strictly related to your Task. So it's up to you. It would be great if you could do the easy ones, but if it starts being tricky (like for how `smartapp.handleHttpCallback` sends the request off to `signature.signatureIsVerified` for even more processing), then don't bother with those ones.

File: index.js
Comment (trisapeace) [inline index.js:374]: This function also uses `req.body.signal`

File: index.js
Comment (trisapeace) [inline index.js:636]: Thank you for removing unnecessary newlines!

File: index.js
Comment (trisapeace) [inline index.js:329]: Could you please add some tests to show that these Validators work as expected?

File: index.js
Comment (trisapeace) [inline index.js:326]: The current `if`-statement in `app.get('/logout'` can probably be replaced by a Validator check.

The current `if`-statment in `app.get('/dashboard'` can probably be replaced by a Validator check.

`app.get('/dashboard/:locationId'` could use the Validator to check for `req.session.user` and `req.cookies.user_sid`. No need for it to validate `locationId` though since we wouldn't get into this route without one.

**Review by trisapeace — COMMENTED** [review]
On the right track! Thanks for doing this :smile:

File: index.js
Comment (risatronic) [reply index.js:347]: Definitely something I can look into, but yeah as you point out I'm not sure I want to mess around with functions from other libraries where stuff is obfuscated like in `handleHttpCallback`

File: index.js
Comment (risatronic) [reply index.js:329]: I have a question about this in the dev channel because I am not sure what the header authorisation token for smartapp is or how to find it, it gets to that function though meaning it passes the validator which is also in that message so I'm not sure if that would be testing enough to check?

https://bravely.slack.com/archives/CE32NB85N/p1610578347334700

File: index.js
Comment (risatronic) [reply index.js:326]: I only applied it to post routes as I wasn't sure - if I do that does `sessionChecker` on line 89 still need to exist or could I also just use that on `app.get('/'`?

File: db/db.js
Comment (risatronic) [reply db/db.js:13]: Sure can! :) I also just noticed that the env file @mariocimet shared with me doesn't actually include `$PG_CONNECTION_STRING` but it's mentioned in the readme, and I don't know that it's actually being used anywhere

Comment (risatronic) [reply db/db.js:13]: It is used after `-d` in `setup_postgresql_local.sh` though and the current env info/template is in 1password so I would also need to update that to be clear we'd need a `$PG_DATABASE` environment variable 🤔

Comment (risatronic) [reply db/db.js:13]: there is no `.env.example` file as far as I know is why

File: index.js
Comment (risatronic) [reply index.js:336]: I'm not sure, I will set it off in postman and see 😅 for this one the error that comes back is re smartapp looking for an authorisation header that I don't have so I didn't notice it as it doesn't make it to the `else`, but this came from where `Validator` is used in Buttons so I think it might be needed 🤔 I'll check though

Comment (risatronic) [reply index.js:336]: ```2021-01-15T00:14:07.178Z - Bad request, parameters missing {"errors":[{"msg":"Invalid value","param":"lifecycle","location":"body"}]}```

Comment (risatronic) [reply index.js:336]: I think missing is more descriptive than invalid value personally, especially because lifecycle wants to be `===` to something and it being wrong is different than not existing at all, but am happy to do whatever!

Comment (risatronic) [reply index.js:336]: (that's what `.exists()` comes back with)

File: index.js
Comment (risatronic) [reply index.js:326]: line *94 I was looking at a different branch locally

File: index.js
Comment (risatronic) [reply index.js:347]: Also this lines up with the linting because it doesn't like that random object that is just `= req.body` and then most of what's inside it goes unused

File: db/db.js
Comment (trisapeace) [reply db/db.js:13]: `$PG_CONNECTION_STRING` was removed by @mariocimet  when he added in `PG_HOST`, `PG_USER`, etc. I guess he forgot to update the README.

Yup, you can update the `.env` in 1Password to include it. It can have extra things that aren't used yet, so you can add it there now so that you don't forget later if you want.

File: index.js
Comment (trisapeace) [reply index.js:336]: Yup, looks like they purposely are not super explicit in their messages:
https://express-validator.github.io/docs/custom-error-messages.html

File: index.js
Comment (trisapeace) [reply index.js:329]: Can you stub out the smartapp function that looks for the token?

File: index.js
Comment (trisapeace) [reply index.js:326]: Interesting. Good point. I hadn't looked at `sessionChecker` too closely. It actually seems to me like it's doing something strange. I would have expected it to redirect to `/login` if the session was missing and otherwise to go on to `next()`. If I'm understanding this correctly. If it did that, then it could also be used in `/dashboard` and `/dashboard/:locationId`... I think.

**Comment by trisapeace** [comment]
Also, yes, it's probably worth waiting to also add this to Sampath and Mario's PR once it's merged. I think they are close.

File: index.js
Comment (risatronic) [reply index.js:349]: @trisapeace can you please confirm that what we'd be wanting in this case would be something like

```
if(validationErrors.isEmpty()){
    const { locationid, state, rpm, distance, mov_f, mov_s } = req.body

    redis.addXeThruSensorData(locationid, state, rpm, distance, mov_f, mov_s)
    handleSensorRequest(locationid)

    res.status(200).send()
} 
```

File: db/db.js
Comment (risatronic) [reply db/db.js:13]: My current plan for this is to be converting it back to `PG_USER` before merging because I'm not sure when the best time to update that across things will be, if anyone thinks it should actually be now though I can try to figure that out. I kept forgetting the other day when pushing linting things and it was breaking Travis 😩

Comment (risatronic) [reply db/db.js:13]: nvm, it has now been added to the `.travis.yml` file and is in the `.env` on 1password per discussion earlier today :) and after running the build only 4 times it decided to pass with that change 🤨 Better than an automatic break though I guess haha

Comment (risatronic) [reply db/db.js:13]: Will update other instances as needed 👍

File: index.js
Comment (risatronic) [reply index.js]: ^I'm just looking at this again now, did that seem to be a better idea than validator for those get requests? If so I can shuffle it around, I think it's kinda left over from when `/dashboard` didn't exist, it says middleware by it though so maybe it can just go away 🤔

Comment (risatronic) [reply index.js:328]: @trisapeace @mariocimet  I finally have the rest of things working so was wondering about this?

File: index.js
Comment (risatronic) [reply index.js:331]: Is it just this function you were wanting added tests for? If the express thing doesn't work the existing tests fail is how it's otherwise working

**Comment by risatronic** [comment]
this is driving me literally insane because I can't tell if tests are janky because of what I did or...? They all pass locally so I can't replicate it on my machine, or at least how to do that isn't immediately apparent. The most times of running them it's taken before to get a pass is 4, I've done it like 10+ times now and always some combo of the last few is failing so if anyone has any insight I'd greatly appreciate it 😭

**Comment by risatronic** [comment]
@mariocimet

@trisapeace took a look at failing Travis builds with me and suggested adding awaits into the api/xethru endpoint (more info here: https://app.clickup.com/t/hx0jjc?comment=168274643) which worked to make it happy but we weren't sure if that would be ok for real life situations, ie. whether we should leave this as updated or amend the tests

I seem to remember a conversation around experimenting with sleeps for the tests also but couldn't find a task for it so I'm not sure one was made, if it was then this could possibly resolve that 🤔

File: index.js
Comment (risatronic) [reply index.js:321]: I have added tests for this but Samsung itself throws an error on top of returning a 401 (which the test does look for and gets), is that a problem? I'm not sure how to get around it

File: index.js
Comment (risatronic) [inline index.js:102]: having this the other way around said too many redirects when I tested it, so did a negative check instead

File: index.js
Comment (risatronic) [inline index.js:107]: updated this because with sessionChecker in place, if someone is logged in and tries to go to /, they should be redirected to the landing page vs login page

File: index.js
Comment (risatronic) [inline index.js:111]: removed sessionChecker from this because it makes no sense with new logic

File: index.js
Comment (mariocimet) [inline index.js:378]: as noted in our call could be a good idea to check for the fields we use that it gets from request.body.data

Comment (risatronic) [reply index.js]: Updated :)

Comment (mariocimet) [reply index.js]: looks good! this is good to go then

Comment (mariocimet) [reply index.js]: I think I'd prefer a squash commit with a nice message? @trisapeace you agree?

**Review by mariocimet — APPROVED** [review]

File: index.js
Comment (risatronic) [reply index.js]: Haha yeah that was definitely my plan cause it's a bit of a mess on that front

**Comment by trisapeace** [comment]
@risatronic Could you please resolve the conflicts with the base branch so that I can see exactly what will end up being merged? Thanks!

**Comment by risatronic** [comment]
> @risatronic Could you please resolve the conflicts with the base branch so that I can see exactly what will end up being merged? Thanks!

can do, I didn't realise there were conflicts but I guess that makes sense with the deployment & rebase haha

**Comment by risatronic** [comment]
@trisapeace done :)

File: .travis.yml
Comment (trisapeace) [inline .travis.yml:41]: After your merge, the `PG_DATABASE` value is in here twice.

File: index.js
Comment (trisapeace) [inline index.js:102]: I think that this should be an OR `||` instead of an AND `&&`. Since if they are missing either of these things, we want them to sign in again.

In general, the opposite of `A && B` is `!A || !B`.

File: index.js
Comment (trisapeace) [reply index.js:106]: Thanks for these explanatory notes :smile:

File: index.js
Comment (trisapeace) [inline index.js:347]: Looks like there used to be a `.json("OK")` instead of `send()` for this response before. It probably doesn't matter, but to be on the safe side, you could add it back in.

File: index.js
Comment (trisapeace) [inline index.js:367]: This also used to have a `.json("OK")` instead of `send()`

File: index.js
Comment (trisapeace) [inline index.js:378]: Where are `data.*.deviceid', 'data.*.data', 'data.*.control'` being used?

I see `request.data.data` on line 391, but don't see the others or the reason for the stars. Can you please help me understand?

File: index.js
Comment (trisapeace) [inline index.js:421]: Looks like the Validator could also check for `data`, `data.device.network.signal.strength` and `data.device.cloud.disconnects`. Or was this left out on purpose?

File: index.js
Comment (trisapeace) [reply index.js:321]: Looks like you figured it out. Thanks for adding the tests. They look good.

File: index.js
Comment (risatronic) [reply index.js:102]: I had actually made a note to ask about that and then forgot 😅

File: index.js
Comment (risatronic) [reply index.js:378]: This came from a discussion with Mario about checking the content of data which is an object, this is the syntax for checking object keys within body for the express validator :)

File: index.js
Comment (risatronic) [reply index.js:421]: Not purposefully left out that I know of, @mariocimet only mentioned checking inside data for the purpose of the one route where I have `data.*.thing` in it but that's easy to add!

File: index.js
Comment (risatronic) [reply index.js:347]: In addition to send or as a replacement? Can do either way :)

File: .travis.yml
Comment (risatronic) [reply .travis.yml:41]: Will fix that, thanks!

File: index.js
Comment (trisapeace) [reply index.js:347]: It looks like our other functions are using just `json("OK")`. I guess it doesn't need a `send()` if you do that :shrug:

File: index.js
Comment (trisapeace) [reply index.js:378]: That makes sense for `request.data.data`, but I just don't see where we are using the other object keys.

File: index.js
Comment (risatronic) [reply index.js:347]: Okay sounds good haha

File: index.js
Comment (risatronic) [reply index.js:378]: @mariocimet mentioned to look inside data for I think `deviceid` specifically so I checked what we were using in the tests to determine what a valid request would entail was my thinking

Comment (trisapeace) [reply index.js:378]: @mariocimet , it doesn't look like our code would break if the door didn't send us the deviceid (or the others). What are you hoping to catch with this?

Comment (risatronic) [reply index.js:378]: Yeah now that it's mentioned isn't deviceid one of the columns we were wanting to remove too? 😅 I think it was more a point of consistency in any case re checking for things not explicitly mentioned in the route handlers

File: index.js
Comment (risatronic) [reply index.js:421]: In this case would I need to explicitly check for data as if it didn't exist the other two wouldn't either?

File: index.js
Comment (risatronic) [reply index.js:347]: Oh, now that I'm looking at this in my text editor I remember why I did this, it was for consistency since the other things for 400 and 500 are formatted this way in the else & catch

```
} else {
  helpers.log(`Bad request, parameters missing ${JSON.stringify(validationErrors)}`)
  res.status(400).send()
  }
} catch (err) {
   helpers.log(err)
   res.status(500).send()
}
```

File: index.js
Comment (risatronic) [reply index.js:421]: mm ok, I am looking at this too and those fields aren't included in the tests so it would make tests fail and I'm not sure what the values of them are supposed to be to be valid if I were to add them 🤔

from `serverTest.js`:

```
it('should return 400 to a device vitals signal with an unregistered coreID', async () => {
  let response = await chai.request(server).post('/api/devicevitals').send({coreid: 'unregisteredID',data:{deviceid:'5A:2B:3C',data: 'closed',control: 'AA'}})
  expect(response).to.have.status(400)
})
```

File: index.js
Comment (trisapeace) [reply index.js:347]: That also makes sense. I'd just say that taking out the `.json("OK")` will require more testing because you've changed the conversation between the devices. You're choice :smile:

File: index.js
Comment (trisapeace) [reply index.js:378]: I'd prefer the philosophy that we want to be as strict as we need to be and no stricter. If they want to change their API to return different values, we should only complain if they were values that we actually care about.

File: index.js
Comment (trisapeace) [reply index.js:421]: To find out some valid values, you could try asking @sampathsatti . Or just try it out - you could add a log message, deploy to dev, and then watch what comes in.

Comment (trisapeace) [reply index.js:421]: It would also be a good opportunity to write some tests that will fail when those values aren't provided.

File: index.js
Comment (risatronic) [reply index.js:347]: Yeah I will probably just put it back for now just on the off chance 😅

File: index.js
Comment (mariocimet) [reply index.js:421]: yeah it's just data.data

Comment (mariocimet) [reply index.js:421]: those values are in the IM21 Door Status enum, we already vary it in the tests using the enum for open vs. closed signals

`'use strict'

const IM21_DOOR_STATE = {
    OPEN: '02',
    CLOSED: '00',
    HEARTBEAT_OPEN: '0A',
    HEARTBEAT_CLOSED: '08',
    LOW_BATT: '04'
}`

File: index.js
Comment (mariocimet) [reply index.js:378]: I was just referring to the request.data.data bit which we actually use to tell us the door status.

Comment (risatronic) [reply index.js:378]: oh ok, so should I change it just to look for that one?

Comment (risatronic) [reply index.js:378]: My question more specifically too I think is that you can't look at `request.data` as far as I can see from the validator documentation, which is why I was looking for the data.data bit of `request.body`, is that incorrect?

"Creates a validation chain for one or more fields. They may be located in any of the following request objects:

req.body
req.cookies
req.headers
req.params
req.query"

**Review by trisapeace — APPROVED** [review]
Approving this with the assumption that the changes that we just went over in review/pairing with Mario will be included. Good job!

**Pull request merged**
